src/instructlab/sdg/utils/chunkers.py

```diff
@@ -0,0 +1,563 @@
+        document_chunks = []
+        prev_page_number = None
+        book_title = None
+
+        for idx, book_element in enumerate(json_book["main-text"]):
+            if book_element["type"] in [
+                "page-footer",
+                "picture",
+                "reference",
+                "meta-data",
+                "figure",
+                "page-header",
+            ]:
+                continue
+            elif book_element["type"] == "footnote":
+                current_book_page_number = book_element["prov"][0]["page"]
+            elif book_element["type"] in [
+                "subtitle-level-1",
+                "paragraph",
+                "table",
+                "title",
+                "equation",
+            ]:
+                if book_element["type"] == "table":
+                    current_book_page_number = self.get_table_page_number(json_book, idx)
+                else:
+                    current_book_page_number = book_element["prov"][0]["page"]
+                    book_text = book_element["text"]
+
+                if book_element["type"] == "subtitle-level-1":
+                    if book_title is None:
+                        book_title = book_text
+                        book_text = f"# Title: **{book_text}**"
+                    else:
+                        book_text = f"## **{book_text}**"
+
+                if book_element["type"] == "title":
+                    book_text = f"# **{book_text}**"
+                if book_element["type"] == "page-header":
+                    book_text = f"Page Header: **{book_text}**\n\n"
+
+                if chunking_criteria is not None:
+                    # custom break function that can be used to chunk document
+                    if chunking_criteria(book_text):
+                        document_chunks.append("\n\n".join(current_buffer))
+                        current_buffer = []
+                elif (
+                    prev_page_number is not None
+                    and prev_page_number != current_book_page_number
+                ) and keep_same_page_thing_together:
+                    document_chunks.append("\n\n".join(current_buffer))
+                    current_buffer = []
+                else:
+                    if (
+                        self.get_token_count("\n\n".join(current_buffer), tokenizer)
+                        >= max_token_per_chunk
+                        and len(current_buffer) > 1
+                    ):
+                        document_chunks.append("\n\n".join(current_buffer[:-1]))
+
+                        if (
+                            self.get_token_count(current_buffer[-1], tokenizer)
+                            >= max_token_per_chunk
+                        ):
+                            document_chunks.append(current_buffer[-1])
+                            current_buffer = []
+                        else:
+                            current_buffer = current_buffer[-1:]
```

Comment:
I can't make any sense of what this code is intended to do if it hits a book_text whose size is greater than max_token_per_chunk.  It looks to me like it just wipes out everything in the buffer _except_ the item that is too big and then continues?  I don't get it.